src/openfermion/utils/_grid_test.py

```diff
@@ -57,3 +57,29 @@ def test_properties(self):
             (2, 1),
             (2, 2),
         ])
+
+    def test_repr(self):
+        self.assertEqual(
+            repr(Grid(dimensions=5, length=3, scale=0.5)),
+            'Grid(dimensions=5, length=3, scale=0.5)')
+
+    def test_str(self):
+        self.assertEqual(
+            str(Grid(dimensions=5, length=4, scale=0.5)),
+            '[-0.25, -0.125, 0.0, 0.125]**5')
+
+    def test_equality(self):
+        g = Grid(dimensions=5, length=3, scale=0.5)
+        g2 = Grid(dimensions=5, length=3, scale=0.5)
+
+        self.assertTrue(g == g2)
+        self.assertFalse(g != g2)
+        self.assertEqual(g, g2)
+        self.assertEqual(hash(g), hash(g2))
+
+        self.assertFalse(g == Grid(dimensions=4, length=3, scale=0.5))
+        self.assertFalse(g == Grid(dimensions=5, length=2, scale=0.5))
+        self.assertFalse(g == Grid(dimensions=5, length=3, scale=0.75))
+
+        self.assertTrue(g != '')
```

Comment:
`self.assertNotEqual()` will work here :)

Reply:
`assertNotEqual` and `assertEqual` assume the class has properly-functioning equality. You have to be careful to avoid using them (or only use them sparingly) when testing equality in the first place.

For example, do you know whether `assertNotEqual` calls `==` or `!=` internally? Or maybe it calls the `__eq__` method on both sides and check for consistency? And regardless of what the code says, what does the library *guarantee*?

In this case, I'm making sure that the correct behavior occurs when an incompatible type `==`'d against Grid. Replacing it with `assertEqual` is not guaranteed to test the same thing.

Reply:
Can you provide a case / instance where this isn't true? The docs for unittest explicitly say "This method should also be avoided when more specific methods are available (e.g. assertEqual(a, b) instead of assertTrue(a == b)), because they provide a better error message in case of failure."

Reply:
Sure:

```python
    def test_tricky_equality_typo(self):
        class C():
            def __init__(self, x):
                self.x = x

            def __eq__(self, other):
                return self.x == other.x

            def __ne__(self, other):
                # Oops: I accidentally wrote self instead of other
                return self.x == self.x

        self.assertEqual(C(1), C(1))
        self.assertNotEqual(C(1), C(2))
```

It is impossible to write a test that will detect the typo using `assertEqual` and `assertNotEqual`.

... you know what, it makes sense to add an EqualsTester class along with this. Hold on...